docs/src/userguide/nogil.rst

```diff
@@ -0,0 +1,160 @@
+
+If neither of these two points apply then you probably do not need to release the GIL. The sort of Cython 
+code that can run without the GIL (no calls to Python, purely C-level numeric operations)
+is often the sort of code that runs efficiently. This sometimes gives people the impression that the
+inverse is true and the trick is releasing the GIL, rather than the actual code they’re running.
+Don’t be misled by this -- your (single-threaded) code will run the same speed with or without the GIL.
+
+Marking functions as able to run without the GIL
+------------------------------------------------
+
+You can mark a whole function (either a Cython function or an :ref:`external function <nogil>`) as
+``nogil`` by appending this to the function signature or by using the ``@cython.nogil`` decorator:
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+    
+        .. code-block:: python
+
+            @cython.nogil(True)
+            @cython.cfunc
+            @cython.noexcept
+            def some_func() -> None:
+            ...
+
+    .. group-tab:: Cython
+    
+        .. code-block:: cython
+    
+            cdef void some_func() noexcept nogil:
+                ....
+
+Be aware that this does not release the GIL when calling the function. It merely indicates that
+a function is suitable for use when the GIL is released. It is also fine to call these functions
+while holding the GIL.
+
+In this case we've marked the function as ``noexcept`` to indicate that it cannot raise a Python
+exception. Be aware that a function with an ``except *`` exception specification (typically functions
+returning ``void``) will be expensive to call because Cython will need to temporarily reacquire
+the GIL after every call to check the exception state. Most other exception specifications are
+cheap to handle in a ``nogil`` block since the GIL is only acquired if an exception is
+actually thrown.
+
+Releasing (and reacquiring) the GIL
+-----------------------------------
+
+To actually release the GIL you can use context managers
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+    
+        .. code-block:: python
+        
+            with cython.nogil(True):
+                ...  # some code that runs without the GIL
+                with cython.gil:
+                    ...  # some code that runs with the GIL
+                ...  # some more code without the GIL
+            
+    .. group-tab:: Cython
+    
+        .. code-block:: cython
+    
+            with nogil:
+                ...  # some code that runs without the GIL
+                with gil:
+                    ...  # some code that runs with the GIL
+                ...  # some more code without the GIL
+            
+The ``with gil`` block is a useful trick to allow a small
+chunk of Python code to in a non-GIL block. Try not to use it
+too much since there is a cost to acquiring the GIL, and because these
+blocks cannot be parallelized.
```

Comment:
```suggestion
too much since there is a cost to waiting for and acquiring the GIL, and because these
blocks cannot run in parallel since all executions require the same lock.
```